### ISSUE TITLE
Scans terminate immediately upon IO errors even when `--allow-io-errors` is enabled

### ISSUE THREAD

**Issue by copylefty** [wordfence/wordfence-cli#109]
This is specific to our environment, but running a scan fails on symlinks to files within an LXC container. EG:

```
$ wordfence malware-scan --allow-io-errors --output-format csv --output-path /root/wordfence/malware_scan_$(date +%Y-%m-%d-%H-%M-%S).csv /data/
...
...
Processing file: /data/04/7674/html/wp-content/cache/autoptimize/index.html
Processing file: /data/04/7674/html/wp-content/cache/autoptimize/js/autoptimize_d41d8cd98f00b204e9800998ecf8427e.js
Processing file: /data/04/7674/html/wp-content/cache/autoptimize/js/autoptimize_fallback.js
Processing file: /data/04/7674/html/wp-content/cache/autoptimize/js/index.html
Exception occurred during scanning: An exception occurred in a child process: Directory search of /data/04/7674/html/wp-content failed ([Errno 2] No such file or directory: '/data/04/7674/html/wp-content/db.php')
1 timeout(s) occurred during scan
```

The error occurs because a user has installed the Query Monitor plugin which symlinks `/data/04/7674/html/wp-content/db.php` to its local mount: 

```
ls -alh /data/04/7674/html/wp-content/db.php
lrwxrwxrwx 1 webuser www-data 64 Aug  9 11:27 /data/04/7674/html/wp-content/db.php -> /var/www/html/wp-content/plugins/query-monitor/wp-content/db.php
```

We're running wordfence-cli from a dedicated node that is just scanning /data/, and it can't find `/var/www/html/` because it's not running within the LXC container. 

Basically from the wordfence-cli end I was hoping for an option to simply skip files that it can't find (`No such file or directory`). Presumably this doesn't count as an IO error and isn't skipped when `--allow-io-errors` is used.

Similarly it fails when it comes across strange filenames:

```
Exception occurred during scanning: An exception occurred in a child process: Directory search of /data/06/8832/html/wp-content/<xmpG:groupName>Grises</xmpG:groupName>
                  <xmpG:groupType>1</xmpG:groupType>
                  <xmpG:Colorants>
                     <rdf:Seq>
```

```
ls /data/06/8832/html/wp-content/
'[24-Feb-2021 17:00:19 UTC] PHP Warning:  include('
 ai1wm-backups
 backupwordpress-b09a546349-backups
 debug.log
'<div class="wrap advanced-excerpt">'$'\r\n\t''<h2><?php _e( "Advanced Excerpt Options", '\''advanced-excerpt'\'' ); ?><'
'Files placed in this directory are automatically included by WordPress MU.'$'\n''There'\''s no need to activate them on a plugins page.'$'\n''tinymce_982d78a0a085795c3235d1d8a6f787ee.gz'
 index.php
 mu-pluginsKEEP-16-Nov-2021
'ns#"><rdf:Description rdf:about="" xmlns:xmp="http:'
'# Only allow direct access to specific Web-available files.'$'\n\n''# Apache 2.2'$'\n''<IfModule !mod_authz_core.c>'$'\n\t''Order Deny,Allow'$'\n\t''Deny from all'$'\n''<'
'%PDF-1.5'$'\r''%'$'\342\343\317\323\r\n''404 0 obj'$'\r''<<'
'<?php'$'\n'
'<?php'$'\r\n'
'<?php'$'\r\n\r\n'
'<?php'$'\r\n''class Advanced_Excerpt {'$'\r\n\t''public $options;'$'\r\n\r\n\t'
'<?php'$'\n\n''function ldap_addmenuuser() {'$'\n''        $objCurrUser = wp_get_current_user();'$'\n''        $objUser = wp_cache_get($objCurrUser->id, '\''users'\'');'$'\n\n''        if (function_exists('\''add_submenu_page'\'')) {'$'\n''                '
 plugins
'p-settings.php on line 84'$'\n''[21-Sep-2021 15:36:41 UTC] PHP Warning:  include('
 themes
 upgrade
 uploads
 webtoffee_export
 wptouch-data
'<xmpG:groupName>Grises<'
'<xmpG:type>PROCESS<'
'<?xpacket end="w"?>'$'\r\n''endstream'$'\r''endobj'$'\r''403 0 obj'$'\r''<<'
'ype>'$'\n''                           <xmpG:red>232<
```

Obviously that needs resolving and it's not surprising WF couldn't read them. But again we'd like wordfence-cli just to skip (and maybe log an error) rather than halting the scan when it comes across things like this

**Comment by akenion** [comment]
With `--allow-io-errors` specified (which is the default behavior as of v2.0.1), scans should continue if the errors above are encountered.

From your first example, I see that there is a `1 timeout(s) occurred during scan` message after the exception. This is only logged if the scan completes without error. Are you also seeing the `Found {X} suspicious files...` message or is it terminating after that?

The `Exception occurred during scanning: ` message should just be a warning when `--allow-io-errors` is enabled and the scan should complete successfully. If it's logging that and then terminating, that's a bug and we will need to investigate further.

**Comment by copylefty** [comment]
Thanks @akenion 

It looks like a bug in this case.

Scanning a 1.6TB directory it's stopping the scan here:

```
Processing file: /data/05/10951/html/wp-content/uploads/sites/5/backup/revslider/aviation-creatives-header/index.php
Processing file: /data/05/10951/html/wp-content/uploads/sites/5/backup/revslider/index.php
File added to scan queue: /data/05/10951/html/wp-content/uploads/sites/5/backup/revslider/slider-1/index.php
Processing file: /data/05/10951/html/wp-content/uploads/sites/5/backup/revslider/slider-1/index.php
Exception occurred during scanning: An exception occurred in a child process: Directory search of /data/05/10951/html/wp-content/uploads/sites/5/bb-platform-previews/43ebd646a98dab403e0fe006ad430d68 failed ([Errno 2] No such file or directory: '/data/05/10951/html/wp-content/uploads/sites/5/bb-platform-previews/43ebd646a98dab403e0fe006ad430d68/0ec5dabe26cd50f8a2c8adec999ab8b2')
2 timeout(s) occurred during scan
Found 2024 suspicious file(s) after processing 109806 file(s) containing 1.6 GiB over 365 second(s)
```

**Comment by akenion** [comment]
The fact that it displays the `Found 2024 suspicious files` message indicates that the scan ran to completion, as does the fact that it reports having scanned 1.6 GiB (assuming you mean 1.6 GB in your post). What are you seeing that indicates otherwise?

**Comment by akenion** [comment]
Alternately, if the scanned directory is in fact 1.6TB, note that Wordfence CLI only scans files with certain extensions by default (such as `.php`). If you want to scan everything, you can use the `--include-all-files` (or `-a`) flag. Specifying patterns and/or exact filenames to include/exclude is also supported.

Also note that only the first 50 MiB of each file is scanned by default. This can be adjusted with the `--scanned-content-limit` (`-M`) option. Also note that by default (unless `--match-all` is specified), Wordfence CLI will stop scanning a file after any content matching a malware sample is found, so any data beyond that will not be read and hence will not be included in the count.

Depending on the options used, it's expected that the amount of files and data scanned will be less than what's actually present in the target directory.

**Comment by copylefty** [comment]
Yes it's 1.6TB

I've removed the folder that gave an error above and re-scanned. It now gets further but again halts when it comes across an error:

```
Processing file: /data/17/12188/html/wp-admin/widgets-form.php
Processing file: /data/17/12188/html/wp-admin/widgets.php
Processing file: /data/17/12188/html/wp-blog-header.php
Processing file: /data/17/12188/html/wp-comments-post.php
Processing file: /data/17/12188/html/wp-config-sample.php
Processing file: /data/17/12188/html/wp-config.php
Exception occurred during scanning: An exception occurred in a child process: Directory search of /data/17/12188/html/wp-content failed ([Errno 2] No such file or directory: '/data/17/12188/html/wp-content/db.php')
6 timeout(s) occurred during scan
Found 2143 suspicious file(s) after processing 438747 file(s) containing 10.4 GiB over 1632 second(s)
```

This does seem like a bug, with `--allow-io-errors

**Comment by akenion** [comment]
Thanks @copylefty. You are correct, this is a bug and I was able to recreate it. I will look into this and implement a fix soon.

**Comment by ewodrich** [comment]
Verified malware scans are completing when various IO errors are encountered during the scan including removing files, renaming directories in various places within the scan path, and file permissions issues. Variations on the malware-scan tests include the following: 

- with (allow-io-errors), without (no-allow-io-errors), and undefined (defaults to allow-io-errors)
- verbose as default or --no-verbose
- routing to output path or not defining an output path
- stderr logging or undefined
- regular user or root
- includes scanning various symlink setups
- verified file counts match between total files in scan path vs the totals of files scanned and files skipped due to file not found when renaming a directory in the scan path
ex. total files: 41391, renamed directory: 2480, malware scanner results: Found 25907 suspicious file(s) after processing 38911 file(s) containing 2.0 GiB over 309 second(s)

**Comment by akenion** [comment]
@copylefty This issue should be resolved with v2.0.3.
